app/controllers/exercise_solutions_controller.rb

```diff
@@ -4,9 +4,16 @@ class ExerciseSolutionsController < ApplicationController
   before_action :authenticate!
 
   def create
+    guide_previously_done = guide_done?
     @solution = @exercise.submit_solution(current_user, solution_params)
     @solution.run_tests!
-    render partial: 'exercise_solutions/results', locals: {solution: @solution}
+    guide_finished_by_solution = !guide_previously_done && guide_done?
+    render partial: 'exercise_solutions/results', locals: {solution: @solution, guide_finished_by_solution: guide_finished_by_solution}
+  end
+
+  def guide_done?
+    guide = @exercise.guide
```

Comment:
Also, it could be a exercise instance method like `exercise.guide_done?`. That way we could test it.

(Sorry if you get this message twice, but Github deleted mu previous one on last commit)